planner/funcdep/fd_graph.go

```diff
@@ -549,6 +549,280 @@ func (s *FDSet) MakeCartesianProduct(rhs *FDSet) {
+		// This is because that the outer join would generate null-extended rows. So if at least one row from the left side
+		// is not null. We can guarantee that the there's no same part between the original rows and the generated rows.
+		// So the null extended rows would not break the original functional dependency.
+		if edge.from.Intersects(innerFDs.NotNullCols) {
+			// One of determinant are not null column, strict FD are kept.
+			// According knowledge #2, we can't take use of right filter's not null attribute.
+			s.addFunctionalDependency(edge.from, edge.to, edge.strict, edge.equiv)
+		} else {
+			// Otherwise, the strict FD are downgraded to a lax one.
+			s.addFunctionalDependency(edge.from, edge.to, false, edge.equiv)
+		}
+	}
+	leftCombinedFDFrom := NewFastIntSet()
+	leftCombinedFDTo := NewFastIntSet()
+	for _, edge := range filterFDs.fdEdges {
+		// Rule #3.2, constant FD are removed from right side of left join.
+		if edge.isConstant() {
+			s.Rule333Equiv.Edges = append(s.Rule333Equiv.Edges, &fdEdge{
+				from:   edge.from,
+				to:     edge.to,
+				strict: edge.strict,
+				equiv:  edge.equiv,
+			})
+			s.Rule333Equiv.InnerCols = innerCols
+			continue
+		}
+		// Rule #3.3, we only keep the lax FD from right side pointing the left side.
+		if edge.equiv {
+			// equivalence: {superset} --> {superset}, either `from` or `to` side is ok here.
+			// Rule 3.3.1
+			if !opt.SkipFDRule331 {
+				leftCover := edge.from.Intersection(outerCols)
+				rightCover := edge.from.Intersection(innerCols)
+				if leftCover.Len() > 0 && rightCover.Len() > 0 {
+					if opt.TypeFDRule331 == CombinedFD {
+						leftCombinedFDFrom.UnionWith(leftCover)
+						leftCombinedFDTo.UnionWith(rightCover)
+					} else {
+						// singleFD case.
+						s.addFunctionalDependency(leftCover, rightCover, true, false)
+					}
+				}
+			}
+
+			// Rule 3.3.2
+			equivColsRight := edge.from.Intersection(innerCols)
+			equivColsLeft := edge.from.Intersection(outerCols)
+			rightAllCols := copyRightFDSet.AllCols()
+			leftAllCols := copyLeftFDSet.AllCols()
+			coveringStrictKeyRight := rightAllCols.SubsetOf(copyRightFDSet.ClosureOfStrict(equivColsRight))
+			coveringStrictKeyLeft := leftAllCols.SubsetOf(copyLeftFDSet.closureOfStrict(equivColsLeft))
+			if coveringStrictKeyLeft && coveringStrictKeyRight {
+				// find the minimum strict Key set, and add
+				s.addFunctionalDependency(copyLeftFDSet.ReduceCols(equivColsLeft), rightAllCols.Union(leftAllCols), true, false)
+			}
+
+			// Rule 3.3.3
+			// need to break down the superset of equivalence, adding each lax FD of them.
+			laxFDFrom := equivColsRight
+			laxFDTo := equivColsLeft
+			for i, ok := laxFDFrom.Next(0); ok; i, ok = laxFDFrom.Next(i + 1) {
+				for j, ok := laxFDTo.Next(0); ok; j, ok = laxFDTo.Next(j + 1) {
+					s.addFunctionalDependency(NewFastIntSet(i), NewFastIntSet(j), false, false)
```

Comment:
What's happening here? We are adding a lax FD from every column in the `from` to every column in the `to`?

Reply:
Yes, the strict equiv is not correct anymore. We would downgrade it to a bunch of lax ones.

Reply:
I mean, shouldn't we add {laxFDFrom} ~> {laxFDTo}?
I think we can only derive one lax FD from one equiv.

Reply:
Got it. I misunderstand it before.